### ISSUE TITLE
Problem passing nested form dict to test request.

### ISSUE THREAD

**Issue by toonarmycaptain** [pallets/werkzeug#1646]
I have a nested form (generated from a series of subforms) generated by Flask_WTF, such that my form can be accessed in my route as `first = form.form_field_1.data` and `second = form.nested_form.nested_form_field_1.data`
Note that this works and does what I expect it to when running regularly. 

My problem is faking this out in a request inside pytest to test the route/data placed into the session cookie: 
First using original dict:
```
request_data = {'form_field_1': 'whatever data',
                'nested_form': {'nested_form_field_1': 'whatever more data'},
                }
```
and in my test using `response = test_client.post('myroute/', data=request_data)`

Results in an error:
```
self = <flask.testing.EnvironBuilder object at 0x00000162924A8588>
key = 'nested_form', value = {'nested_form_field_1': 'whatever more data'}

    def _add_file_from_data(self, key, value):
        """Called in the EnvironBuilder to add files from the data dict."""
        if isinstance(value, tuple):
            self.files.add_file(key, *value)
        elif isinstance(value, dict):
            from warnings import warn
    
            warn(
                "Passing a dict as file data is deprecated as of"
                " version 0.5 and will be removed in version 1.0. Use"
                " a tuple or 'FileStorage' object instead.",
                DeprecationWarning,
                stacklevel=2,
            )
            value = dict(value)
            mimetype = value.pop("mimetype", None)
            if mimetype is not None:
                value["content_type"] = mimetype
>           self.files.add_file(key, **value)
E           TypeError: add_file() got an unexpected keyword argument 'nested_form_field_1'

..\..\my_mcve_env\lib\site-packages\werkzeug\test.py:474: TypeError
```

If in the route I print my `request.form` to the console (when running regularly outside testing) I get: `ImmutableMultiDict([('csrf_token', '...'), ('form_field_1', 'whatever data'), ('nested_form-csrf_token',
 '...'), ('nested_form-nested_form_field_1', 'whatever more data'), ('submit', 'Submit')])`

And I see the dict has a dash between the nested form name and the field name, so I think, maybe I can fake it out since it looks like it's flattening the dict, and I pass:
```
request_data = {'form_field_1': 'whatever data',
                'nested_form-nested_form_field_1': 'whatever more data',
                }
```

Which results in the error:
```
 assert flask.session['nested_form-nested_form_field_1'] == {
                'nested_form_field_1': request_data['nested_form']['nested_form_field_1']}
E           KeyError: 'nested_form'
```
even though the request.form in my print debug statement reads (I disable CSRF for testing):
```
ImmutableMultiDict([('form_field_1', 'whatever data',), ('nested_form-nested_form_field_1', 'whatever more data')])
```

Now when I try this second method, comment out the code that would use the nested form field, and disable the assert that tests the nested value is passed to the session, the test passes, so I know the first field is passing fine in the test. 

I found somewhere to try passing the form dict as json:
```
response = test_client.post(
            'myroute/', data=json.dumps(request_data), content_type='application/json')
```
But this doesn't seem to pass even the first field to the 
```
>           assert flask.session['form_field_1'] == request_data['form_field_1']

my_mcve_app.py:101: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\my_mcve_env\lib\site-packages\werkzeug\local.py:378: in <lambda>
    __getitem__ = lambda x, i: x._get_current_object()[i]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <SecureCookieSession {}>, key = 'form_field_1'

    def __getitem__(self, key):
        self.accessed = True
>       return super(SecureCookieSession, self).__getitem__(key)
E       KeyError: 'form_field_1'

..\..\my_mcve_env\lib\site-packages\flask\sessions.py:84: KeyError
```

It doesn't appear to place anything inside the form key inside the request, but when I instead print the request.__dict__ in my app during the test I get:
```
{'environ': {'REMOTE_ADDR': '127.0.0.1', 'HTTP_USER_AGENT': 'werkzeug/0.15.6', 'REQUEST_METHOD': 'POST', 'SCRIPT_NAME': '', 'PATH_INFO': 'myroute/', 'QUERY_STRING': '', 'REQUEST_URI': 'myroute/', 'RAW_URI': 'myroute/', 'SERVER_NAME': 'localhost', 'SERVER_PORT': '80', 'HTTP_HOST': 'localhost', 'SERVER_PROTOCOL': 'HTTP/1.1', 'wsgi.version': (1, 0), 'wsgi.url_scheme': 'http', 'wsgi.input': <_io.BytesIO object at 0x000001E747453C50>, 'wsgi.errors': <_pytest.capture.EncodedFile object at 0x000001E74643F550>, 'wsgi.multithread': False, 'wsgi.multiprocess': False, 'wsgi.run_once': False, 'CONTENT_TYPE': 'application/json', 'CONTENT_LENGTH': '72', 'HTTP_CONTENT_TYPE': 'application/json', 'HTTP_CONTENT_LENGTH': '72', 'flask._preserve_context': True, 'werkzeug.request': <Request 'http://localhost/myroute/' [POST]>}, 'shallow': False, 'cookies': {}, 'url_rule': <Rule '/myroute/' (OPTIONS, GET, POST, HEAD) -> my_mcve_app.my_mcve>, 'view_args': {}, 'stream': <werkzeug.wsgi.LimitedStream object at 0x000001E74743F630>, 'form': ImmutableMultiDict([]), 'files': ImmutableMultiDict([]), '_parsed_content_type': ('application/json', {}), '_cached_data': b'{'form_field_1': 'whatever data', 'nested_form': {'nested_form_field_1': 'whatever more data'}}', '_cached_json': ({'form_field_1': 'whatever data', 'nested_form: {'nested_form_field_1': 'whatever more data'}}, {'form_field_1': 'whatever data', 'nested_form': {'nested_form_field_1': 'whatever more data'}}), 'url': 'http://localhost/myroute/'}
```

I comprehend the warning that shows passing dicts is deprecated and will be removed in 1.0, but this is werkzeug v0.15, flask v1.1.1, pytest v5.1.2, so I would not expect it to be removed yet. 
Either way - since this is the production behaviour - nesting dicts with flask-wtf is possible, which leads to nested dicts in the request, unless I'm doing something wrong that I would be better achieving another way?

**Comment by davidism** [comment]
Form data is not nestable, it's always sent as a flat list of keys/values. WTForms accomplishes nesting by applying prefixes to the keys so that it can reconstruct the nesting after loading the flat data. You'll need to provide this same flat data.

**Comment by toonarmycaptain** [comment]
You know what, I feel pretty dumb now. I figured out that although I was flattening the data using ImmutableMultiDict, I was still trying to refer to that flattened data using dict syntax. 
Obviously `ImmutableMultiDict([('form_field_1', 'whatever data',), ('nested_form-nested_form_field_1', 'whatever more data')])` doesn't have a key `nested_form`. I fixed this by flattening the input dict in test set up, then comparing with the same dict in my assertion.
